Sources/NIOEmbedded/AsyncEmbeddedEventLoop.swift

```diff
@@ -0,0 +1,357 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2022 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+#if compiler(>=5.5.2) && canImport(_Concurrency)
+import Dispatch
+import _NIODataStructures
+import NIOCore
+import NIOConcurrencyHelpers
+
+
+/// An `EventLoop` that is thread safe and whose execution is fully controlled
+/// by the user.
+///
+/// Unlike more complex `EventLoop`s, such as `SelectableEventLoop`, the `AsyncEmbeddedEventLoop`
+/// has no proper eventing mechanism. Instead, reads and writes are fully controlled by the
+/// entity that instantiates the `AsyncEmbeddedEventLoop`. This property makes `AsyncEmbeddedEventLoop`
+/// of limited use for many application purposes, but highly valuable for testing and other
+/// kinds of mocking. Unlike `EmbeddedEventLoop`, `AsyncEmbeddedEventLoop` is fully thread-safe and
+/// safe to use from within a Swift concurrency context.
+///
+/// Unlike `EmbeddedEventLoop`, `AsyncEmbeddedEventLoop` does require that user tests appropriately
+/// enforce thread safety. Used carefully it is possible to safely operate the event loop without
+/// explicit synchronization, but it is recommended to use `establishContext` in any case where it's
+/// necessary to ensure that the event loop is not making progress.
+///
+/// Time is controllable on an `AsyncEmbeddedEventLoop`. It begins at `NIODeadline.uptimeNanoseconds(0)`
+/// and may be advanced by a fixed amount by using `advanceTime(by:)`, or advanced to a point in
+/// time with `advanceTime(to:)`.
+///
+/// If users wish to perform multiple tasks at once on an `AsyncEmbeddedEventLoop`, it is recommended that they
+/// use `establishContext` to perform the operations.
+@available(macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0, *)
+public final class AsyncEmbeddedEventLoop: EventLoop, @unchecked Sendable {
```

Comment:
Generally speaking I think we should treat `@unchcked Sendable` as 'unsafe': we should comment why it's okay in this instance.

Reply:
Should this be `NIOAsyncEmbeddedEventLoop`?